src/Sulu/Component/HttpCache/EventSubscriber/InvalidationSubscriber.php

```diff
@@ -57,25 +91,100 @@ public function __construct(
     public static function getSubscribedEvents()
     {
         return [
-            Events::PUBLISH => ['invalidateDocumentForPublishing', -512],
+            Events::PUBLISH => ['invalidateDocumentBeforePublishing', 600],
+            Events::REMOVE => ['invalidateDocumentBeforeRemoving', 600],
         ];
     }
 
-    public function invalidateDocumentForPublishing(PublishEvent $event)
+    /**
+     * Invalidates all urls in in the locale of the document of already published documents when they
+     * get republished (eg on content change).
+     * This method is executed before the actual publishing of the document.
+     *
+     * @param PublishEvent $event
+     */
+    public function invalidateDocumentBeforePublishing(PublishEvent $event)
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof ResourceSegmentBehavior || !$document->getPublished()) {
+            // no need to invalidate the path to the document if it is not published currently
+            return;
+        }
+
+        // invalidate urls of the locale of the document
+        $locale = $this->documentInspector->getLocale($document);
+        foreach ($this->getLocaleUrls($document, $locale) as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
+
+    /**
+     * Invalidates all urls in all locales of the document whey they get removed.
+     * This method is executed before the actual removing of the document.
+     *
+     * @param RemoveEvent $event
+     */
+    public function invalidateDocumentBeforeRemoving(RemoveEvent $event)
     {
         $document = $event->getDocument();
 
-        if (!$document instanceof StructureBehavior) {
+        if (!$document instanceof ResourceSegmentBehavior) {
             return;
         }
 
-        $structureBridge = $this->structureManager->wrapStructure(
-            $this->documentInspector->getMetadata($document)->getAlias(),
-            $this->documentInspector->getStructureMetadata($document)
-        );
+        // invalidate urls of all locales of the document
+        $urls = [];
+        foreach ($this->documentInspector->getLocales($document) as $locale) {
+            $urls = array_merge($urls, $this->getLocaleUrls($document, $locale));
+        }
+
+        foreach ($urls as $url) {
```

Comment:
I don't think that you need two loops here... Why don't you just write `$this->pathHandler->invalidatePath($url)` in the upper loop? The URLs will always be different, since different localizations of a page can't have the same URL, so the `array_merge` is also not necessary.